zebrad/tests/common/get_block_template_rpcs/submit_block.rs

```diff
@@ -1,170 +1,88 @@
-//         (past the MAX_BLOCK_REORG_HEIGHT) via getblock rpc calls, get the finalized tip height
-//         of the updated cached state, restart zebra without peers, and submit blocks above the
-//         finalized tip height.
-
-use std::path::PathBuf;
+//! This test requires a cached chain state that is partially synchronized close to the
+//! network chain tip height. It will finish the sync and update the cached chain state.
+//!
+//! After finishing the sync, it will get the first 20 blocks in the non-finalized state
+//! (past the MAX_BLOCK_REORG_HEIGHT) via getblock rpc calls, get the finalized tip height
+//! of the updated cached state, restart zebra without peers, and submit blocks above the
+//! finalized tip height.
 
-use color_eyre::eyre::{eyre, Context, Result};
+use color_eyre::eyre::{Context, Result};
 
-use futures::TryFutureExt;
-use indexmap::IndexSet;
 use reqwest::Client;
-use tower::{Service, ServiceExt};
-use zebra_chain::{block::Height, parameters::Network, serialization::ZcashSerialize};
-use zebra_state::HashOrHeight;
-use zebra_test::args;
+use zebra_chain::parameters::Network;
 
 use crate::common::{
-    cached_state::{copy_state_directory, start_state_service_with_cache_dir},
-    config::{persistent_test_config, testdir},
-    launch::ZebradTestDirExt,
-    lightwalletd::random_known_rpc_port_config,
+    cached_state::get_raw_future_blocks,
+    launch::{can_spawn_zebrad_for_rpc, spawn_zebrad_for_rpc},
+    test_type::TestType,
 };
 
-use super::cached_state::{load_tip_height_from_state_directory, ZEBRA_CACHED_STATE_DIR};
-
-async fn get_future_block_hex_data(
-    network: Network,
-    zebrad_state_path: &PathBuf,
-) -> Result<Option<String>> {
-    tracing::info!(
-        ?zebrad_state_path,
-        "getting cached sync height from ZEBRA_CACHED_STATE_DIR path"
-    );
+#[allow(clippy::print_stderr)]
+pub(crate) async fn run() -> Result<()> {
+    let _init_guard = zebra_test::init();
 
-    let cached_sync_height =
-        load_tip_height_from_state_directory(network, zebrad_state_path.as_ref()).await?;
+    // We want a zebra state dir in place,
+    let test_type = TestType::UpdateZebraCachedState;
+    let test_name = "submit_block_test";
+    let network = Network::Mainnet;
 
-    let future_block_height = Height(cached_sync_height.0 + 1);
+    // Skip the test unless the user specifically asked for it and there is a zebrad_state_path
+    if !can_spawn_zebrad_for_rpc(test_name, test_type) {
+        return Ok(());
+    }
 
     tracing::info!(
-        ?cached_sync_height,
-        ?future_block_height,
-        "got cached sync height, copying state dir to tempdir"
+        ?network,
+        ?test_type,
+        "running submitblock test using zebrad",
     );
 
-    let copied_state_path = copy_state_directory(network, &zebrad_state_path).await?;
+    let raw_blocks: Vec<String> = get_raw_future_blocks(network, test_type, test_name, 3).await?;
```

Comment:
Nitpick: let's use a constant so it is easy to change if the test fails

Reply:
Good idea, updated in [`d72ea40`](https://github.com/ZcashFoundation/zebra/pull/5589/commits/d72ea40650530d1dfa5e3bdd73d9d146eb28fba8)